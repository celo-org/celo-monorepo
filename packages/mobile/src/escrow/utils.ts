import { ensureLeading0x } from '@celo/base'
import { generateDeterministicInviteCode } from '@celo/utils/lib/account'
import { publicKeyToAddress } from '@celo/utils/lib/address'

export const generateEscrowPaymentIdAndPk = (
  recipientPhoneHash: string,
  recipientPepper: string,
  addressIndex: number = 0
) => {
  const { publicKey, privateKey } = generateDeterministicInviteCode(
    recipientPhoneHash,
    recipientPepper,
    addressIndex
  )
  return { paymentId: publicKeyToAddress(publicKey), privateKey: ensureLeading0x(privateKey) }
}

export const generateUniquePaymentId = (
  existingPaymentIds: string[],
  phoneHash: string,
  pepper: string
) => {
  const paymentIdSet: Set<string> = new Set(existingPaymentIds)

  // Using an upper bound of 100 to be sure this doesn't run forever given
  // the realistic amount of pending escrow txs is far less than this
  for (let i = 0; i < 100; i += 1) {
    const { paymentId } = generateEscrowPaymentIdAndPk(phoneHash, pepper, i)
    if (!paymentIdSet.has(paymentId)) {
      return paymentId
    }
  }
}
