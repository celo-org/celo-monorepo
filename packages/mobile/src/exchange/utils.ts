import BigNumber from 'bignumber.js'
import { CURRENCY_ENUM } from 'src/geth/consts'
import { convertToContractDecimals } from 'src/tokens/saga'
import { getContractKitAsync } from 'src/web3/contracts'

export async function celoTocUsd(celoAmount: BigNumber) {
  const celoAmountInWei = (
    await convertToContractDecimals(celoAmount, CURRENCY_ENUM.GOLD)
  ).integerValue()

  const contractKit = await getContractKitAsync()
  const exchange = await contractKit.contracts.getExchange()
  const exchangeRate = await exchange.getGoldExchangeRate(celoAmountInWei)
  const cUsdAmount = exchangeRate.multipliedBy(celoAmount)

  return cUsdAmount
}
