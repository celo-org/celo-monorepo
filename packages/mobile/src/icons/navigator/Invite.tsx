import * as React from 'react'
import Svg, { ClipPath, Defs, G, Path } from 'react-native-svg'

export function Invite() {
  return (
    <Svg width={32} height={32} viewBox="0 0 32 32" fill="none">
      <G clipPath="url(#prefix__clip0)">
        <Path
          d="M22.564 12.903l2.775 2.501A2 2 0 0126 16.89V26a2 2 0 01-2 2H7a2 2 0 01-2-2v-9.047a2 2 0 01.74-1.553l3.078-2.497"
          stroke="#B4B9BD"
          strokeWidth={2}
          strokeLinecap="round"
        />
        <Path
          d="M26 17l-6.204 4m-8.878 0L5 17"
          stroke="#B4B9BD"
          strokeWidth={2}
          strokeLinecap="round"
          strokeLinejoin="round"
        />
        <Path
          d="M5.247 26.81c-.37.409-.343 1.05.058 1.43a.97.97 0 001.395-.05l-1.453-1.38zm19.533 1.378a.97.97 0 001.395.055 1.03 1.03 0 00.046-1.43l-1.441 1.375zM18.736 21l.006 1.013L18.736 21zm-6.038 0l-.006-1.013.006 1.013zm-2.166-.046l-5.285 5.855L6.7 28.19l5.285-5.855-1.453-1.381zm2.173 1.059h6.037l-.013-2.026h-6.037l.013 2.026zm6.764.325l5.31 5.85 1.442-1.376-5.311-5.85-1.441 1.376zm-.727-.325c.275 0 .539.117.727.325l1.44-1.376a2.947 2.947 0 00-2.18-.975l.013 2.026zm-6.757.322a.97.97 0 01.72-.322l-.013-2.026c-.822 0-1.604.35-2.16.967l1.453 1.38z"
          fill="#B4B9BD"
        />
        <Path
          d="M13.103 9.348v-.002a8.487 8.487 0 011.739-3.656c.8-.966 1.73-1.533 2.528-1.672l.035-.006.024-.006a.773.773 0 01.071 0h.029l.028-.002a2.097 2.097 0 011.701.685c.665.743 1.005 2.004.715 3.591h0l-.001.006c-.245 1.384-.856 2.654-1.657 3.609-.807.961-1.751 1.543-2.642 1.687-.78.118-1.423-.1-1.885-.595h0l-.007-.007c-.687-.721-1.02-2.007-.678-3.632z"
          stroke="#B4B9BD"
          strokeWidth={2}
        />
      </G>
      <Defs>
        <ClipPath id="prefix__clip0">
          <Path d="M0 0h32v32H0V0z" fill="#fff" />
        </ClipPath>
      </Defs>
    </Svg>
  )
}

export default React.memo(Invite)
