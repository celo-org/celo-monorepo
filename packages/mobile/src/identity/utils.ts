import BigNumber from 'bignumber.js'

export function isUserBalanceSufficient(
  userBalance: string | number | null,
  estimatedTxAmount: number | string
) {
  if (!userBalance || new BigNumber(userBalance).isLessThan(new BigNumber(estimatedTxAmount))) {
    return false
  }

  return true
}
