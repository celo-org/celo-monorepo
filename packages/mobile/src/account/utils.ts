import { ADDRESS_LENGTH } from 'src/exchange/reducer'

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}
