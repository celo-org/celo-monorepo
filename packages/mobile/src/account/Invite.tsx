import TextInput, { TextInputProps } from '@celo/react-components/components/TextInput'
import withTextInputLabeling from '@celo/react-components/components/WithTextInputLabeling'
import colors from '@celo/react-components/styles/colors'
import * as React from 'react'
import { WithTranslation } from 'react-i18next'
import { StyleSheet, View } from 'react-native'
import SafeAreaView from 'react-native-safe-area-view'
import { NavigationInjectedProps } from 'react-navigation'
import { connect } from 'react-redux'
import { defaultCountryCodeSelector } from 'src/account/reducer'
import { hideAlert, showError } from 'src/alert/actions'
import CeloAnalytics from 'src/analytics/CeloAnalytics'
import { CustomEventNames } from 'src/analytics/constants'
import { componentWithAnalytics } from 'src/analytics/wrapper'
import { ErrorMessages } from 'src/app/ErrorMessages'
import i18n, { Namespaces, withTranslation } from 'src/i18n'
import ContactPermission from 'src/icons/ContactPermission'
import Search from 'src/icons/Search'
import { importContacts } from 'src/identity/actions'
import { e164NumberToAddressSelector, E164NumberToAddressType } from 'src/identity/reducer'
import { headerWithCancelButton } from 'src/navigator/Headers'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { filterRecipients, NumberToRecipient, Recipient } from 'src/recipients/recipient'
import RecipientPicker from 'src/recipients/RecipientPicker'
import { recipientCacheSelector } from 'src/recipients/reducer'
import { RootState } from 'src/redux/reducers'
import { SendCallToAction } from 'src/send/SendCallToAction'
import { navigateToPhoneSettings } from 'src/utils/linking'
import { requestContactsPermission } from 'src/utils/permissions'

const InviteSearchInput = withTextInputLabeling<TextInputProps>(TextInput)

interface State {
  searchQuery: string
  hasGivenContactPermission: boolean
}

interface Section {
  key: string
  data: Recipient[]
}

interface StateProps {
  defaultCountryCode: string
  e164PhoneNumberAddressMapping: E164NumberToAddressType
  recipientCache: NumberToRecipient
}

interface DispatchProps {
  showError: typeof showError
  hideAlert: typeof hideAlert
  importContacts: typeof importContacts
}

const mapDispatchToProps = {
  showError,
  hideAlert,
  importContacts,
}

type Props = StateProps & DispatchProps & WithTranslation & NavigationInjectedProps

const mapStateToProps = (state: RootState): StateProps => ({
  defaultCountryCode: defaultCountryCodeSelector(state),
  e164PhoneNumberAddressMapping: e164NumberToAddressSelector(state),
  recipientCache: recipientCacheSelector(state),
})

class Invite extends React.Component<Props, State> {
  static navigationOptions = () => ({
    ...headerWithCancelButton,
    headerTitle: i18n.t('sendFlow7:invite'),
  })

  state: State = { searchQuery: '', hasGivenContactPermission: true }

  async componentDidMount() {
    await this.tryImportContacts()
  }

  tryImportContacts = async () => {
    const { recipientCache } = this.props

    // If we've imported already
    if (Object.keys(recipientCache).length) {
      return
    }

    const hasGivenContactPermission = await requestContactsPermission()
    this.setState({ hasGivenContactPermission })

    if (hasGivenContactPermission) {
      this.props.importContacts()
    }
  }

  onSearchQueryChanged = (searchQuery: string) => {
    this.setState({ searchQuery })
  }

  onSelectRecipient = (recipient: Recipient) => {
    this.props.hideAlert()
    if (recipient.e164PhoneNumber) {
      CeloAnalytics.track(CustomEventNames.friend_invited)
      navigate(Screens.InviteReview, { recipient })
    } else {
      this.props.showError(ErrorMessages.CANT_SELECT_INVALID_PHONE)
    }
  }

  onPressContactsSettings = () => {
    navigateToPhoneSettings()
  }

  buildSections = (): Section[] => {
    const { t, recipientCache } = this.props
    // Only recipients without an address are invitable
    const invitableRecipients = Object.values(recipientCache)
      .filter((val) => !val.address)
      .sort(({ displayName: a }, { displayName: b }) => (b > a ? -1 : 1))

    const queryRecipients = (recipients: Recipient[]) =>
      filterRecipients(recipients, this.state.searchQuery)

    const sectionInputs = [{ title: t('contacts'), recipients: Object.values(invitableRecipients) }]
    return sectionInputs
      .map((input) => ({
        key: input.title,
        data: queryRecipients(input.recipients),
      }))
      .filter((section) => section.data.length > 0)
  }
  renderListHeader = () => {
    const { t } = this.props
    const { hasGivenContactPermission } = this.state

    if (hasGivenContactPermission) {
      return null
    }

    return (
      <SendCallToAction
        icon={<ContactPermission />}
        header={t('importContactsCta.header')}
        body={t('importContactsCta.body')}
        cta={t('importContactsCta.cta')}
        onPressCta={this.onPressContactsSettings}
      />
    )
  }

  render() {
    return (
      <SafeAreaView style={style.container}>
        <View style={style.textInputContainer}>
          <InviteSearchInput
            value={this.state.searchQuery}
            onChangeText={this.onSearchQueryChanged}
            icon={<Search />}
            placeholder={this.props.t('nameOrPhoneNumber')}
          />
        </View>
        <RecipientPicker
          testID={'RecipientPicker'}
          sections={this.buildSections()}
          searchQuery={this.state.searchQuery}
          defaultCountryCode={this.props.defaultCountryCode}
          onSelectRecipient={this.onSelectRecipient}
          listHeaderComponent={this.renderListHeader}
        />
      </SafeAreaView>
    )
  }
}

const style = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.background,
  },
  textInputContainer: {
    paddingBottom: 5,
    borderBottomColor: colors.listBorder,
    borderBottomWidth: 1,
  },
})

export default componentWithAnalytics(
  connect<StateProps, DispatchProps, {}, RootState>(
    mapStateToProps,
    mapDispatchToProps
  )(withTranslation(Namespaces.sendFlow7)(Invite))
)
