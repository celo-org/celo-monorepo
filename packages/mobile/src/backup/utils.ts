import { sampleSize } from 'lodash'
import { AsyncStorage } from 'react-native'
import { generateMnemonic, wordlists } from 'react-native-bip39'
import { getKey, setKey } from 'src/utils/keyStore'
import Logger from 'src/utils/Logger'

const TAG = 'Backup/utils'

export async function createQuizWordList(mnemonic: string, language: string | null) {
  const disallowedWordSet = new Set(mnemonic.split(' '))
  const languageWordList = getWordlist(language)
  const wordOptions: string = await generateMnemonic(1000, null, languageWordList)
  const quizWordList = new Set(
    [...wordOptions.split(' ')].filter((word: string) => !disallowedWordSet.has(word))
  )
  return [...quizWordList]
}

export function selectQuizWordOptions(correctWord: string, allWords: string[], numOptions: number) {
  const wordOptions = []
  const correctWordPosition = Math.floor(Math.random() * numOptions)
  const randomWordIndexList = sampleSize([...Array(allWords.length).keys()], numOptions - 1)
  let randomWordIndex: number = 0

  for (let i = 0; i < numOptions; i++) {
    if (i === correctWordPosition) {
      wordOptions.push(correctWord)
      continue
    }

    wordOptions.push(allWords[randomWordIndexList[randomWordIndex]])
    randomWordIndex += 1
  }
  return wordOptions
}

export function getWordlist(language: string | null) {
  if (!language) {
    return wordlists.EN
  }

  switch (language.slice(0, 2)) {
    case 'es': {
      return wordlists.ES
    }
    default: {
      return wordlists.EN
    }
  }
}

// TODO(Rossy) Remove after the next alfa testnet reset
export async function getStoredMnemonic(): Promise<string | null> {
  try {
    Logger.debug(TAG, 'Checking keystore for mnemonic')
    let mnemonic = await getKey('mnemonic')
    if (mnemonic) {
      return mnemonic
    }

    Logger.debug(TAG, 'Mnemonic not found in keystore, checking async storage')
    mnemonic = await AsyncStorage.getItem('mnemonic')
    if (mnemonic) {
      await setKey('mnemonic', mnemonic)
      await AsyncStorage.removeItem('mnemonic')
      return mnemonic
    }

    Logger.error(TAG, 'No mnemonic found')
    return null
  } catch (error) {
    Logger.error(TAG, 'Failed to retrieve mnemonic', error)
    return null
  }
}
