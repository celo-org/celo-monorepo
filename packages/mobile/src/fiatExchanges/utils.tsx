import { SIMPLEX_URI } from 'src/config'
import { CURRENCY_ENUM } from 'src/geth/consts'
import { LocalCurrencyCode } from 'src/localCurrency/consts'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { navigateToURI } from 'src/utils/linking'

export const openMoonpay = (currencyCode: LocalCurrencyCode, currencyToBuy: CURRENCY_ENUM) => {
  navigate(Screens.MoonPay, {
    localAmount: 0,
    currencyCode,
    currencyToBuy,
  })
}

export const openSimplex = (account: string | null) => {
  navigateToURI(`${SIMPLEX_URI}?address=${account}`)
}
