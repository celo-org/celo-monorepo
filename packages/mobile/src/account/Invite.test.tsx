import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import * as renderer from 'react-test-renderer'
import Invite from 'src/account/Invite'
import { createMockStore } from 'test/utils'
import { mockE164NumberToInvitableRecipient, mockNavigation } from 'test/values'

describe('Invite', () => {
  it('renders correctly with recipients', () => {
    const tree = renderer.create(
      <Provider
        store={createMockStore({
          recipients: { recipientCache: mockE164NumberToInvitableRecipient },
        })}
      >
        {/*
          // @ts-ignore */}
        <Invite navigation={mockNavigation} />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with no recipients', () => {
    const tree = renderer.create(
      <Provider store={createMockStore({})}>
        {/*
          // @ts-ignore */}
        <Invite navigation={mockNavigation} />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })
})
