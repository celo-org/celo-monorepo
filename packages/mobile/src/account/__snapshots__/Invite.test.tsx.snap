// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invite renders correctly with no recipients 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderColor": "#EEEEEE",
            "borderRadius": 3,
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 50,
            "marginHorizontal": 10,
            "marginVertical": 5,
          },
          Object {
            "borderColor": "#EEEEEE",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "center",
            "minWidth": 45,
          }
        }
      >
        <svg
          fill="none"
          height={18}
          style={Object {}}
          viewBox="0 0 18 18"
          width={18}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M2.65971 7.31402C2.65971 9.88456 4.74354 11.9684 7.31408 11.9684C9.88462 11.9684 11.9685 9.88456 11.9685 7.31402C11.9685 4.74348 9.88462 2.65964 7.31408 2.65964C4.74354 2.65964 2.65971 4.74348 2.65971 7.31402ZM6.10352e-05 7.31402C6.10352e-05 11.3534 3.27466 14.628 7.31408 14.628C8.65119 14.628 9.9045 14.2692 10.983 13.6427L14.9508 17.6105C15.4701 18.1298 16.3121 18.1298 16.8315 17.6105C17.3508 17.0912 17.3508 16.2492 16.8315 15.7298L13.0074 11.9058C14.0211 10.6505 14.6281 9.05315 14.6281 7.31402C14.6281 3.2746 11.3535 0 7.31408 0C3.27466 0 6.10352e-05 3.2746 6.10352e-05 7.31402Z"
            fill="#81868B"
            fillRule="evenodd"
            style={Object {}}
          />
        </svg>
      </View>
      <View
        style={
          Array [
            Object {
              "height": 35,
              "width": 1,
            },
            Object {
              "backgroundColor": "#EEEEEE",
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
            },
            Object {
              "flex": 1,
              "height": 54,
              "marginHorizontal": 8,
            },
          ]
        }
      >
        <TextInput
          allowFontScaling={true}
          keyboardType="default"
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="nameOrPhoneNumber"
          placeholderTextColor="rgba(0, 0, 0, .4)"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "fontFamily": "Hind-Regular",
              },
              Object {
                "borderColor": "#D1D5D8",
                "borderRadius": 3,
                "padding": 8,
              },
              Object {
                "backgroundColor": "#FFFFFF",
                "flex": 1,
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
    <RCTScrollView
      ListEmptyComponent={
        <View
          style={
            Object {
              "justifyContent": "center",
              "paddingHorizontal": 50,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "center",
                "paddingVertical": 20,
                "textAlign": "center",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Hind-Regular",
                  "fontSize": 16,
                  "lineHeight": 24,
                }
              }
            >
              noContacts
            </Text>
          </View>
        </View>
      }
      data={Array []}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={30}
      keyExtractor={[Function]}
      keyboardShouldPersistTaps="handled"
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      renderItem={[Function]}
      renderSectionHeader={[Function]}
      scrollEventThrottle={50}
      sections={Array []}
      stickyHeaderIndices={Array []}
      stickySectionHeadersEnabled={false}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={
            Object {
              "justifyContent": "center",
              "paddingHorizontal": 50,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "center",
                "paddingVertical": 20,
                "textAlign": "center",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Hind-Regular",
                  "fontSize": 16,
                  "lineHeight": 24,
                }
              }
            >
              noContacts
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;

exports[`Invite renders correctly with recipients 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderColor": "#EEEEEE",
            "borderRadius": 3,
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 50,
            "marginHorizontal": 10,
            "marginVertical": 5,
          },
          Object {
            "borderColor": "#EEEEEE",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "center",
            "minWidth": 45,
          }
        }
      >
        <svg
          fill="none"
          height={18}
          style={Object {}}
          viewBox="0 0 18 18"
          width={18}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M2.65971 7.31402C2.65971 9.88456 4.74354 11.9684 7.31408 11.9684C9.88462 11.9684 11.9685 9.88456 11.9685 7.31402C11.9685 4.74348 9.88462 2.65964 7.31408 2.65964C4.74354 2.65964 2.65971 4.74348 2.65971 7.31402ZM6.10352e-05 7.31402C6.10352e-05 11.3534 3.27466 14.628 7.31408 14.628C8.65119 14.628 9.9045 14.2692 10.983 13.6427L14.9508 17.6105C15.4701 18.1298 16.3121 18.1298 16.8315 17.6105C17.3508 17.0912 17.3508 16.2492 16.8315 15.7298L13.0074 11.9058C14.0211 10.6505 14.6281 9.05315 14.6281 7.31402C14.6281 3.2746 11.3535 0 7.31408 0C3.27466 0 6.10352e-05 3.2746 6.10352e-05 7.31402Z"
            fill="#81868B"
            fillRule="evenodd"
            style={Object {}}
          />
        </svg>
      </View>
      <View
        style={
          Array [
            Object {
              "height": 35,
              "width": 1,
            },
            Object {
              "backgroundColor": "#EEEEEE",
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
            },
            Object {
              "flex": 1,
              "height": 54,
              "marginHorizontal": 8,
            },
          ]
        }
      >
        <TextInput
          allowFontScaling={true}
          keyboardType="default"
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="nameOrPhoneNumber"
          placeholderTextColor="rgba(0, 0, 0, .4)"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "fontFamily": "Hind-Regular",
              },
              Object {
                "borderColor": "#D1D5D8",
                "borderRadius": 3,
                "padding": 8,
              },
              Object {
                "backgroundColor": "#FFFFFF",
                "flex": 1,
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
    <RCTScrollView
      ListEmptyComponent={
        <View
          style={
            Object {
              "justifyContent": "center",
              "paddingHorizontal": 50,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "center",
                "paddingVertical": 20,
                "textAlign": "center",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Hind-Regular",
                  "fontSize": 16,
                  "lineHeight": 24,
                }
              }
            >
              noContacts
            </Text>
          </View>
        </View>
      }
      ListFooterComponent={[Function]}
      data={
        Array [
          Object {
            "data": Array [
              Object {
                "contactId": "contactId",
                "displayId": "14155550000",
                "displayName": "John Doe",
                "e164PhoneNumber": "+14155550000",
                "kind": "Contact",
                "phoneNumberLabel": "phoneNumLabel",
              },
            ],
            "key": "contacts",
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={30}
      keyExtractor={[Function]}
      keyboardShouldPersistTaps="handled"
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      renderItem={[Function]}
      renderSectionHeader={[Function]}
      scrollEventThrottle={50}
      sections={
        Array [
          Object {
            "data": Array [
              Object {
                "contactId": "contactId",
                "displayId": "14155550000",
                "displayName": "John Doe",
                "e164PhoneNumber": "+14155550000",
                "kind": "Contact",
                "phoneNumberLabel": "phoneNumLabel",
              },
            ],
            "key": "contacts",
          },
        ]
      }
      stickyHeaderIndices={Array []}
      stickySectionHeadersEnabled={false}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#F8F8F8",
                "display": "flex",
                "flexDirection": "row",
                "height": 36,
                "justifyContent": "space-between",
                "paddingHorizontal": 10,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#81868B",
                  "fontFamily": "Hind-Medium",
                  "fontSize": 12,
                }
              }
            >
              contacts
            </Text>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "height": 50,
                  "justifyContent": "space-between",
                  "paddingHorizontal": 10,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    Object {
                      "marginRight": 10,
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "justifyContent": "center",
                      },
                      Object {
                        "backgroundColor": "#7AD6FE",
                        "borderRadius": 20,
                        "height": 40,
                        "width": 40,
                      },
                    ]
                  }
                >
                  <Image
                    source={
                      Object {
                        "testUri": "../../../src/images/unknown-user-icon.png",
                      }
                    }
                    style={
                      Object {
                        "alignItems": "center",
                        "height": 40,
                        "justifyContent": "center",
                        "width": 40,
                      }
                    }
                  />
                </View>
              </View>
              <View
                style={
                  Object {
                    "flex": 1,
                    "flexDirection": "row",
                    "paddingRight": 5,
                  }
                }
              >
                <Text
                  ellipsizeMode="tail"
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "color": "#2E3338",
                        "fontFamily": "Hind-SemiBold",
                        "fontSize": 14,
                        "lineHeight": 18,
                      },
                      Object {
                        "lineHeight": 41,
                      },
                    ]
                  }
                >
                  John Doe
                </Text>
              </View>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#2E3338",
                      "fontFamily": "Hind-SemiBold",
                      "fontSize": 14,
                      "lineHeight": 18,
                    },
                    Object {
                      "fontFamily": "Hind-Light",
                    },
                    Object {
                      "lineHeight": 41,
                      "textAlign": "right",
                    },
                  ]
                }
              >
                14155550000
              </Text>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        />
        <View
          onLayout={[Function]}
        >
          <View
            style={
              Object {
                "backgroundColor": "#EEEEEE",
                "height": 1,
                "marginLeft": 60,
              }
            }
          />
          <Text
            style={
              Array [
                Object {
                  "color": "#81868B",
                  "fontFamily": "Hind-Regular",
                  "fontSize": 14,
                  "lineHeight": 20,
                },
                Object {
                  "paddingHorizontal": 50,
                  "paddingVertical": 20,
                  "textAlign": "center",
                },
              ]
            }
          >
            searchFriends
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
